### ISSUE TITLE
SpotArea not being created "SpotArea Cleaning for XX : false"

### ISSUE THREAD

**Issue by Boehmer86** [nicoduj/homebridge-deebotEcovacs#50]
Hello,

First off, congratulation to you plugin and for your effort in supporting it here! It is truly appreciated :-) 

My issue:
- the SpotAreas defined are not getting published

My status:
- my HomeBridge and HomeKit are overall properly configured and working
- your plugin does work normally, sound works
- I used this table to define my Spot area ID's (maybe link it in your Wiki? it is quite helpful) -> https://github.com/mrbungle64/ecovacs-deebot.js/issues/57
- I read this (https://github.com/nicoduj/homebridge-deebotEcovacs/issues/43) bug report and can confirm that I manually used the correct configuration file format 
- I tried using just one ID or different ones to no different results
- I rebooted my RaspberryPi between each config change to update the cache 

My startup log:
[1/10/2021, 6:16:13 PM] Homebridge v1.2.5 is running on port 51374.
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : true
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : true
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - **SpotArea Cleaning for Ozmo : false**
[1/10/2021, 6:16:15 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : false

My configuration:
`        {
            "email": "_removed_",
            "password": "_removed_",
            "countryCode": "DE",
            "deebotNames": [
                "Ozmo"
            ],
            "refreshTimer": 120,
            "publishBipSwitch": true,
            "publishSwitch": true,
            "publishFan": true,
            "publishMotionDetector": true,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "0",
                "1",
                "3,4"
            ],
            "platform": "DeebotEcovacs"
}`

Thanks a lot and many greetings from Berlin

Robert

**Comment by mrbungle64** [comment]
@Boehmer86 

> I used this table to define my Spot area ID's (maybe link it in your Wiki? it is quite helpful) -> mrbungle64/ecovacs-deebot.js#57

Please note: The linked [issue](https://github.com/mrbungle64/ecovacs-deebot.js/issues/57) was marked as "invalid".
This user has posted an excerpt from a library configuration file. We did not understand what he meant by that.

**Comment by Boehmer86** [comment]
Thanks for the hint, I used your reference to confirm that my ID selection is correct based on this https://github.com/mrbungle64/ecovacs-deebot.js/commit/e140f2a9c53aa6d2f6f4eb59389f1f641a8be052#

The issue does still persist of course as I did not change anything

**Comment by julien94800** [comment]
Hello,
Could you please try to remove Deebotnames from your config? Could you confirm that you have only one Deebot?
In my case everything is working and quite same config than you expect that I do not have this variable defined

**Comment by Boehmer86** [comment]
Hi Julien,

Thanks for the quick response!

I updated my config. applied, restarted but the result is sadly the same.

This is "new" current config:
`        {
            "email": "_removed_",
            "password": "_removed_",
            "countryCode": "DE",
            "refreshTimer": 120,
            "publishBipSwitch": true,
            "publishSwitch": false,
            "publishFan": false,
            "publishMotionDetector": false,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "1",
                "3",
                "4",
                "6",
                "12"
            ],
            "platform": "DeebotEcovacs"
        }`

This is the start log:
[1/10/2021, 8:17:23 PM] Homebridge v1.2.5 is running on port 51374.
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : true
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : true
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : false
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : false
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] WARNING - Service Removed00000049-0000-1000-8000-0026BB765291/SwitchOnServiceOzmo/Start/Stop Ozmo
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] WARNING - Service Removed00000040-0000-1000-8000-0026BB765291/FanServiceOzmo/Start/Pause Ozmo
[1/10/2021, 8:17:23 PM] [DeebotEcovacs] WARNING - Service Removed00000085-0000-1000-8000-0026BB765291/MotionServiceOzmo/Ozmo needs attention

With greetings

Robert

**Comment by julien94800** [comment]
Hello,

Other hint, did you try to set the publishswitch to true?
Which model do you have?
From my side it is a 920 and you will find below my settings. For me everything is working


     "countryCode": "FR",
            "refreshTimer": 0,
            "publishBipSwitch": true,
            "publishSwitch": true,
            "publishFan": true,
            "publishMotionDetector": true,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "1",
                "6",
                "7"
            ],
            "platform": "DeebotEcovacs"

**Comment by Boehmer86** [comment]
Dear Julien,

Thanks for sharing your configuration, I tried adjusting mine to yours but with no success sadly.

Here is the latest log:
[1/10/2021, 8:36:11 PM] Homebridge v1.2.5 is running on port 51374.
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : true
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : true
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : false
[1/10/2021, 8:36:12 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : false

Here is my current configuration:
`        {
            "email": "_removed_",
            "password": "_removed_",
            "countryCode": "DE",
            "refreshTimer": 0,
            "publishBipSwitch": true,
            "publishSwitch": true,
            "publishFan": true,
            "publishMotionDetector": true,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "1",
                "3",
                "4",
                "6",
                "12"
            ],
            "platform": "DeebotEcovacs"
        }`

Regarding my Ecovacs Deebot, I am using the latest T8+. It is the same as the T8. Technically there should not be any difference in the API for any of the Deebot robots. 

I will check the source code here briefly, maybe it gives a hint.

Best

Robert

**Comment by julien94800** [comment]
So unfortunately I won't be able to help you more, sorry! 
In the past some differences between models were noticed so that's why I asked but not sure. 
When you modify the config and then restart homebridge, do you see some switches added and or removed depending of your modification?
Did you try to launch homebridge in debug mode?

**Comment by Boehmer86** [comment]
Hi Julien,

I enabled Debug mode and removed a switch which showed in the status log.

This is the debug log:
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - getDeebots : [{"did":"__________-1954-44a5-9039-5e40e650bf2f","name":"E08_____________","class":"fqxoiu","resource":"Cc3g","nick":"Ozmo","company":"eco-ng","bindTs":1609604018847}]
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : true
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : true
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : false
[1/10/2021, 8:56:39 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : false
[1/10/2021, 8:56:40 PM] [DeebotEcovacs] INFO - Vacbot Ozmo ready: "Client connected. Subscribe successful"
...
[1/10/2021, 8:57:10 PM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[1/10/2021, 8:57:10 PM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[1/10/2021, 8:57:10 PM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 
[1/10/2021, 8:57:10 PM] [DeebotEcovacs] INFO - Error from deebot Ozmo : Request Timeout 

I am currently researching where the Request Timeout comes from. 

With greetings

Robert

**Comment by mrbungle64** [comment]
@Boehmer86 

Your model (`"class":"fqxoiu"`) is not supported by version 0.4.26 of ecovacs-deebot library which is still used by the current version of homebridge-deebotEcovacs.

It is supported since version 0.5.0 which was released only some days after 0.4.26

@nicoduj 
Can you please update to the latest version?

**Comment by Boehmer86** [comment]
Hey @mrbungle64 

Thanks, this makes a lot of sense, I checked the ecovacs-deebot library but did not look for the version stated and being used here.

I also further debugged the connection messages and found a more general network problem with the small Netgear Switch connected to my Pi. I am just fixing this one. It should not have any influence on the issue I opened but I am happy to have found it while looking at the debug log. 

I will wait for an update to the library and report back, if there is any way I can test/change something please let me know.

Thanks

Robert

**Comment by nicoduj** [comment]
Hi all, sorry for the delay. Will do the update today I hope !

**Comment by Boehmer86** [comment]
Hi everyone

Big thanks @nicoduj for the very swift update, and thanks to @mrbungle64 and all above this really did the trick, the enumeration works now. I will now experiment with it and see if I can do SpotArea cleaning like I wanted to but the main reason why I opened the thread has been resolved. 

To share the final log after updating:
[1/11/2021, 3:13:01 PM] Homebridge v1.2.5 is running on port 51374.
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - stopping deebots discovery, number of deebots found : 1
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Discovered Deebot : Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Edge Cleaning for Ozmo : false
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Spot Cleaning for Ozmo : false
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - SpotArea Cleaning for Ozmo : true
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - CustomArea Cleaning for Ozmo : true
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating Battery Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating Main Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating Sound stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating Auto stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating SpotArea 0 stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating SpotArea 1 stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating SpotArea 2 stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating SpotArea 3 stateless Switch Service Ozmo
[1/11/2021, 3:13:02 PM] [DeebotEcovacs] INFO - Creating SpotArea 4 stateless Switch Service Ozmo

To share the final config:
`        {
            "email": "_removed_",
            "password": "_removed_",
            "countryCode": "DE",
            "deebotNames": [
                "Ozmo"
            ],
            "refreshTimer": 120,
            "publishBipSwitch": true,
            "publishSwitch": true,
            "publishFan": false,
            "publishMotionDetector": false,
            "publishAutoSwitch": true,
            "publishEdgeSwitch": false,
            "publishSpotSwitch": true,
            "publishSpotAreaSwitches": [
                "1",
                "3",
                "4",
                "6",
                "12"
            ],
            "platform": "DeebotEcovacs"
}`

Interesting side note, I noticed before an elevated CPU usage when having the plugin active, 10-20% this has now dropped to 0-1% again after updating. 

Thanks to you all and with greetings from Berlin

Robert

**Comment by Boehmer86** [comment]
Ok, too confirm and close this case. I just verified it, I can now with Homekit, Siri and my Homepods kindly tell my Ozmo to clean the Kitchen :-) Thanks to Siri Shortcuts any phrase is possible. Mission accomplished, thanks a lot again!

**Comment by nicoduj** [comment]
@Boehmer86 Nice if it works. I am interested into reports on cpu usage, I have such issues myself and can't point if it comes from plugin side, or lib side. It especially cause some hearbeat drift in another plugin, like there was some sync IO somewhere .... @mrbungle64 if you have any idea about that also, you are welcome :)

**Comment by Boehmer86** [comment]
Hey @nicoduj 
So far it keeps working well, I can also confirm that the CPU and memory usage is normal. However you are correct, in my test yesterday I also noticed heartbeat drifts by the Homebridge-Hue plugin due to excessive I/O usage of your plugin. As it is not happening right now this might be either rooted in you updating the library which might have fixed a race condition, or it has to do with me setting the refresh rate to 120 seconds. Before I did not specify it, however the other Homebridge Ecovacs plugin used 120 seconds by default so I figured it might be suitable for me. 
Best
Robert
